src/control/server/instance_storage_rpc.go

```diff
@@ -0,0 +1,215 @@
+	if pciAddr == "" {
+		pciAddr = "<nil>"
+	}
+
+	return &ctlpb.NvmeControllerResult{
+		Pciaddr: pciAddr,
+		State:   newState(log, status, errMsg, infoMsg, "nvme controller "+op),
+	}
+}
+
+// scmFormat will return either successful result or error.
+func (srv *IOServerInstance) scmFormat(p *scm.Provider, reformat bool) (*ctlpb.ScmMountResult, error) {
+	var eMsg, iMsg string
+	srvIdx := srv.Index()
+	cfg := srv.scmConfig()
+	status := ctlpb.ResponseStatus_CTL_SUCCESS
+
+	req, err := scm.CreateFormatRequest(cfg, reformat)
+	if err != nil {
+		return nil, errors.Wrap(err, "generate format request")
+	}
+
+	scmStr := fmt.Sprintf("SCM (%s:%s)", cfg.Class, cfg.MountPoint)
+	srv.log.Infof("Instance %d: starting format of %s", srvIdx, scmStr)
+	res, err := p.Format(*req)
+	if err == nil && !res.Formatted {
+		err = errors.WithMessage(scm.FaultUnknown, "is still unformatted")
+	}
+
+	if err != nil {
+		srv.log.Errorf("  format of %s failed: %s", scmStr, err)
+		return nil, err
+	}
+	srv.log.Infof("Instance %d: finished format of %s", srvIdx, scmStr)
+
+	return newMntRet(srv.log, "format", cfg.MountPoint, status, eMsg, iMsg), nil
+}
+
+func (srv *IOServerInstance) bdevFormat(p *bdev.Provider) (results proto.NvmeControllerResults) {
+	srvIdx := srv.Index()
+	cfg := srv.bdevConfig()
+	results = make(proto.NvmeControllerResults, 0, len(cfg.DeviceList))
+
+	// A config with SCM and no block devices is valid.
+	if len(cfg.DeviceList) == 0 {
+		return
+	}
+
+	bdevListStr := strings.Join(cfg.DeviceList, ",")
+	srv.log.Infof("Instance %d: starting format of %s block devices (%s)",
+		srvIdx, cfg.Class, bdevListStr)
+
+	res, err := p.Format(bdev.FormatRequest{
+		Class:      cfg.Class,
+		DeviceList: cfg.DeviceList,
+	})
+	if err != nil {
+		results = append(results,
+			newCret(srv.log, "format", "", ctlpb.ResponseStatus_CTL_ERR_NVME,
+				err.Error(), fault.ShowResolutionFor(err)))
+		return
+	}
+
+	for dev, status := range res.DeviceResponses {
+		var errMsg, infoMsg string
+		ctlpbStatus := ctlpb.ResponseStatus_CTL_SUCCESS
+		if status.Error != nil {
+			ctlpbStatus = ctlpb.ResponseStatus_CTL_ERR_NVME
+			errMsg = status.Error.Error()
+			srv.log.Errorf("  format of %s device %s failed: %s", cfg.Class, dev, errMsg)
+			if fault.HasResolution(status.Error) {
+				infoMsg = fault.ShowResolutionFor(status.Error)
+			}
+		}
+		results = append(results,
+			newCret(srv.log, "format", dev, ctlpbStatus, errMsg, infoMsg))
+	}
+
+	srv.log.Infof("Instance %d: finished format of %s block devices (%s)",
+		srvIdx, cfg.Class, bdevListStr)
+
+	return
+}
+
+// doFormat performs format on storage subsystems, populates response results
+// in storage subsystem routines and broadcasts (closes channel) if successful.
+func (srv *IOServerInstance) StorageFormat(reformat bool, scmProvider *scm.Provider, bdevProvider *bdev.Provider) (resp *ctlpb.StorageFormatResp) {
```

Comment:
Why are you passing in the scmProvider and bdevProvider instead of using the ones assigned via the instance constructor? https://github.com/daos-stack/daos/blob/master/src/control/server/server.go#L206